traits/trait_dict_object.py

```diff
@@ -62,8 +56,299 @@ def __init__(self, added=None, changed=None, removed=None):
+
+        Parameters
+        ----------
+        adict : dict
+            The dict from which values will be updated into this dict.
+        """
+
+        validated_dict = {}
+        added = {}
+        changed = {}
+
+        for key, value in adict.items():
+            validated_key = self.key_validator(key)
+            validated_value = self.value_validator(value)
+
+            if validated_key in self:
+                changed[validated_key] = self[validated_key]
+            else:
+                added[validated_key] = validated_value
+
+            validated_dict[validated_key] = validated_value
+
+        super().update(validated_dict)
+        self.notify(added=added, changed=changed, removed={})
+
+    def setdefault(self, key, value=None):
+        """ Returns the value if key is present in the dict, else creates the
+        key-value pair and returns the value.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            Key to the item.
+        """
+
+        if key in self:
+            return self[key]
+
+        key = self.key_validator(key)
+        value = self.value_validator(value)
+
+        if key in self:
+            changed = {key: self[key]}
+            added = {}
+        else:
+            changed = {}
+            added = {key: value}
+
+        super().__setitem__(key, value)
+
+        self.notify(added=added, changed=changed, removed={})
+
+        return value
+
+    def pop(self, key, value=Undefined):
+        """ Remove specified key and return the corresponding
+        value. If key is not found, the default value is returned
+        if given, otherwise KeyError is raised.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            Key to the dict item.
+
+        value : any
+            Value to return if key is absent.
+        """
+
+        if value is Undefined or key in self:
+            removed = super().pop(key)
+            self.notify(
+                added={},
+                changed={},
+                removed={key: removed}
+            )
+            return removed
+        return value
+
+    def popitem(self):
+        """ Remove and return some(key, value) pair as a tuple. Raise KeyError
+        if dict is empty.
+
+        Returns
+        -------
+        key_value : tuple
+            Some 2-tuple from the dict.
+
+        Raises
+        ------
+        KeyError
+            If the dict is empty
+        """
+
+        item = super().popitem()
+        self.notify(added={}, changed={}, removed=dict([item]))
+        return item
+
+    # -- pickle and copy support ----------------------------------------------
+
+    def __getstate__(self):
+        """ Get the state of the object for serialization.
+
+        Notifiers are transient and should not be serialized.
+        """
+        result = self.__dict__.copy()
+        # notifiers are transient and should not be serialized
+        result.pop("notifiers", None)
```

Comment:
Why not `del result["notifiers"]`? Can this code ever be called with an object that doesn't have a `notifiers` attribute?

Reply:
Nope, I don't think so,  changed to `del`,